src/auth/aws.js

```diff
@@ -39,12 +40,112 @@ export function createCanonicalRequest(request) {
+}
+
+
+/**
+ * Returns string to sign
+ * 
+ * @export
+ * @param {any} canonicalRequest 
+ * @returns 
+ */
+export function createStringToSign(canonicalRequest, scope) {
+    if (!canonicalRequest) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof canonicalRequest !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+
+    const timeStampISO8601Format = _.getAWSTimestamp();
+    const request = hex(_.toSHA256(canonicalRequest).toString());
+
+    return `${AUTH_IDENTIFIER_HEADER}\n${timeStampISO8601Format}\n${scope}\n${request}`;
+}
+
+/**
+ * Returns Signing key
+ * 
+ * @export
+ * @param {any} awsSecretKey 
+ * @param {any} awsRegion 
+ * @returns {String}
+ */
+export function createSigningKey(awsSecretKey, awsRegion) {
+    if (!awsSecretKey) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof awsSecretKey !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+    if (!awsRegion) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof awsRegion !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+
+    return _.toHmacSHA256(_.toHmacSHA256(_.toHmacSHA256(_.toHmacSHA256(`AWS4${awsSecretKey}`, _.getShortDate()), awsRegion), 's3'), 'aws4_request');
+}
 
+/**
+ * Returns signature
+ * 
+ * @export
+ * @param {any} signingKey 
+ * @param {any} stringToSign 
+ * @returns {String}
+ */
+export function createSignature(signingKey, stringToSign) {
+    if (!signingKey) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof signingKey !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+    if (!stringToSign) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof stringToSign !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+
+    return _.toHmacSHA256(stringToSign, signingKey);
+}
+
+
+/**
+ * Returns Authorization header
+ * 
+ * @export
+ * @param {any} awsAccessKey 
+ * @param {any} awsRegion 
+ * @param {any} signature 
+ * @returns {String}
+ */
+export function createAuthorizationHeader(awsAccessKey, awsRegion, signature) {
+    if (!awsAccessKey) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof awsAccessKey !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+    if (!awsRegion) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof awsRegion !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+    if (!signature) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof signature !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+
+    return `${AUTH_IDENTIFIER_HEADER} Credential=${awsAccessKey}/${_.getShortDate()}/${awsRegion}/s3/aws4_request,SignedHeaders=host;range;x-amz-content-sha256;x-amz-date,Signature=${signature}`;
+}
+
+/**
+ * Returns x-amz-content-sha256 header
+ * 
+ * @export
+ * @returns 
+ */
+export function createXAmzContentSha256Header() {
+    return _.toSHA256('').toString();
+}
+
+
+/**
+ * Returns x-amz-date header
+ * 
+ * @export
+ * @returns 
+ */
+export function createXZmzDateHeader() {
+    return _.getAWSTimestamp();
```

Comment:
I speculate that calling the `getAWSTimestamp` from more than one place will get us into trouble, as the values may occasionally not match. What do you think about calling it from the top level method and then passing it through everywhere else?